artichoke-backend/src/extn/core/array/trampoline.rs

```diff
@@ -142,6 +142,25 @@ pub fn clear(interp: &mut Artichoke, mut ary: Value) -> Result<Value, Error> {
     Ok(ary)
 }
 
+pub fn push<I>(interp: &mut Artichoke, mut ary: Value, others: I) -> Result<Value, Error>
+where
+    I: IntoIterator<Item = Value>,
+    I::IntoIter: Clone,
+{
+    if ary.is_frozen(interp) {
+        return Err(FrozenError::with_message("can't modify frozen Array").into());
+    }
+
+    for value in others {
+        match push_single(interp, ary, value) {
+            Ok(new_ary) => ary = new_ary,
+            Err(e) => return Err(e),
+        }
```

Comment:
Pretty clever idea, however i think it might cause extra/new allocations. `push_single` retrieves the array from memory, adds the item to the address is memory. I think it would be better to bring some of the `push_single` logic into this function. something like:

```rust
let array_mut = unsafe { array.as_inner_mut() };
for value in others {
  array_mut.push(value);
}
```

and then reboxing as per `push_single`.

Reply:
@b-n thanks for the feedback, yea I was just trying to stay away from `unsafe` even though I realize its kinda needed for this stuff.

I tried adjusting based on your suggestion and the way `push_single` was implemented.

Can I ask: conceptually, I understand that `Array` implements the Ruby `Array` API (in Rust), but in terms of the underlying data, how should I think of it? Is it better to perceive it as a pointer-type abstraction? The function names here seem to suggest that its a Box-ed item (e.g. `unbox_from_value` etc.).

Reply:
The Ruby value stored in the array is a tagged union that can either be an integral Ruby type (like nil, true, false, a float, int, or symbol), an interpreter heap-allocated object like a String or Array, or a Rust allocated heap object like a Regexp. These types are stored in a pointer, int64, float64 C union.

The use of "boxed" here is more similar to that of a Java boxed `Integer` than Rust's `Box`.